docs: add v2 to v3 API changes to UPGRADING.md (#2106)
aws/aws-sdk-js-v3: UPGRADING.md

these changes in our fast growing list of [blog posts](https://aws.amazon.com/blogs/developer/category/developer-tools/aws-sdk-for-javascript-in-node-js/). Among the blog posts, The following 2 will get you up-to-speed:

- [modular packages](https://aws.amazon.com/blogs/developer/modular-packages-in-aws-sdk-for-javascript/)
- [middleware stack](https://aws.amazon.com/blogs/developer/first-class-typescript-support-in-modular-aws-sdk-for-javascript/)

Bellow are a summary of interface changes from v2 to v3 SDK. The goal is to help you easily find the v3 equivalents of
the v2 APIs you are already familiar with.

## Client Constructors

This list is indexed by [parameters in v2 SDK](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Config.html). For
options shown as `Planned` in v3, they are confirmed to be supported, but no timeline can be shared at the moment. They
might not have the same name either.

- [`computeChecksums`](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Config.html#computeChecksums-property)
  - **v2**: Whether to compute MD5 checksums for payload bodies when the service accepts it (currently supported in S3
    only).

>   - **v3**: Not available. Planned.

Reviewer: Should we create a GitHub issue for each `Planned` item and link it from UPGRADING.md? This way we will have a GitHub issue which maintainers and users can refer too.

Reviewer: The creation of GitHub issues and linking them in UPGRADING.md are not blocking for this PR, and can be done in a different PR.

- [`convertResponseTypes`](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Config.html#convertResponseTypes-property)
  - **v2**: Whether types are converted when parsing response data.
  - **v3**: **Deprecated**. This option is considered not type-safe because it doesn't convert the types like time stamp
    or base64 binaries from the JSON response.
- [`correctClockSkew`](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Config.html#correctClockSkew-property)
  - **v2**: Whether to apply a clock skew correction and retry requests that fail because of an skewed client
    clock.
  - **v3**: **Deprecated**. SDK _always_ applies a clock skew correction.
- [`systemClockOffset`](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Config.html#systemClockOffset-property)
  - **v2**: An offset value in milliseconds to apply to all signing times.
  - **v3**: No change.
- [`credentials`](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Config.html#credentials-property)
  - **v2**: The AWS credentials to sign requests with.

>   - **v3**: No change. It can also be an async function that returns a credential.

Reviewer: Singular for credentials is credentials, as it has multiple values like accessKeyId, secretAccessKey etc.
```suggestion
  - **v3**: No change. It can also be an async function that returns credentials.
```

>     See [v3 reference](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/interfaces/_aws_sdk_middleware_signing.awsauthinputconfig-1.html#credentials)

Reviewer: ```suggestion
    See [v3 reference for AwsAuthInputConfig credentials](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/interfaces/_aws_sdk_middleware_signing.awsauthinputconfig-1.html#credentials).
```

- [`endpointCacheSize`](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Config.html#endpointCacheSize-property)
  - **v2**: The size of the global cache storing endpoints from endpoint discovery operations.
  - **v3**: Not available. Planned. This option configures endpoint discovery behavior, which is not yet available in v3.
- [`endpointDiscoveryEnabled`](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Config.html#endpointDiscoveryEnabled-property)
  - **v2**: Whether to call operations with endpoints given by service dynamically.
  - **v3**: Not available. Planned. This option configures endpoint discovery behavior, which is not yet available in v3.
- [`hostPrefixEnabled`](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Config.html#hostPrefixEnabled-property)
  - **v2**: Whether to marshal request parameters to the prefix of hostname.
  - **v3**: **Deprecated**. SDK _always_ injects the hostname prefix when necessary.
- [`httpOptions`](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Config.html#httpOptions-property)

>   A set of options to pass to the low-level HTTP request. These options are aggregated differently in V3. You can

Reviewer: ```suggestion
  A set of options to pass to the low-level HTTP request. These options are aggregated differently in v3. You can
```

  configure them by supplying a new `requestHandler`. Here's the example of setting http options in Node.js runtime. You

>   can find more in [v3 reference](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/classes/_aws_sdk_node_http_handler.nodehttphandler-1.html).

Reviewer: ```suggestion
  can find more in [v3 reference for NodeHttpHandler](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/classes/_aws_sdk_node_http_handler.nodehttphandler-1.html).
```

  ```javascript
  const { Agent } = require("https");
  const { Agent: HttpAgnet } = require("http");
  const { NodeHttpHandler } = require("@aws-sdk/node-http-handler");
  const dynamodbClient = new DynamoDBClient({
    requestHandler: new NodeHttpHandler({
      httpsAgent: new Agent({
        /*params*/
      }),
      httpAgent: new HttpAgnet({
        /*params*/
      }),
      connectionTimeout: /*number in milliseconds*/
      socketTimeout: /*number in milliseconds*/
    }),
  });

>   ```

Reviewer: Suggestion: clarify that both httpAgent and httpsAgents are not required. For example:

---

  All v3 requests use HTTPS by default. You only need to provide custom httpsAgent.
  ```javascript
  const { Agent } = require("https");
  const { NodeHttpHandler } = require("@aws-sdk/node-http-handler");

  const dynamodbClient = new DynamoDBClient({
    requestHandler: new NodeHttpHandler({
      httpsAgent: new Agent({
        /*params*/
      }),
      connectionTimeout: /*number in milliseconds*/
      socketTimeout: /*number in milliseconds*/
    }),
  });
  ```

  If you are passing custom endpoint which uses http, then you need to provide httpAgent.
  ```javascript
  const { Agent } = require("http");
  const { NodeHttpHandler } = require("@aws-sdk/node-http-handler");

  const dynamodbClient = new DynamoDBClient({
    requestHandler: new NodeHttpHandler({
      httpAgent: new Agent({
        /*params*/
      }),
    }),
    endpoint: "http://example.com",
  });
  ```

Author: Will do.

>   If the client is suppose to run in browsers, a different set of options is available. You can find more in [v3

Reviewer: ```suggestion
  If the client is running in browsers, a different set of options is available. You can find more in [v3
```

>   reference](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/classes/_aws_sdk_fetch_http_handler.fetchhttphandler-1.html).

Reviewer: ```suggestion
  reference for FetchHttpHandler](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/classes/_aws_sdk_fetch_http_handler.fetchhttphandler-1.html).
```

  ```javascript
  const { FetchHttpHandler } = require("@aws-sdk/fetch-http-handler");
  const dynamodbClient = new DynamoDBClient({
    requestHandler: new FetchHttpHandler({
      requestTimeout: /*number in milliseconds*/
    }),
  });
  ```

>   Each option of `httpOptions` is specified bellow:

Reviewer: ```suggestion
  Each option of `httpOptions` is specified below:
```

>   - `httpOptions.proxy`

Reviewer: ```suggestion
  - `proxy`
```

    - **v2**: The URL to proxy requests through

>     - **v3**: You can set up a proxy with an agent following [this guide](https://docs.aws.amazon.com/sdk-for-javascript/v3/developer-guide/node-configuring-proxies.html)

Reviewer: ```suggestion
    - **v3**: You can set up a proxy with an agent following [Configuring proxies for Node.js](https://docs.aws.amazon.com/sdk-for-javascript/v3/developer-guide/node-configuring-proxies.html)
```

  - `agent`

    - **v2**: The Agent object to perform HTTP requests with. Used for connection pooling.

>     - **v3**: You can configure `httpAgent` or `httpsAgent` like in the examples above.

Reviewer: ```suggestion
    - **v3**: You can configure `httpAgent` or `httpsAgent` as shown in the examples above.
```

  - `xhrAsync`
    - **v2**: Whether the SDK will send asynchronous HTTP requests.
    - **v3**: **Deprecated**. Requests are _always_ asynchronous.
  - `xhrWithCredentials`
    - **v2**: Sets the "withCredentials" property of an XMLHttpRequest object.
    - **v3**: Not available. SDK inherits [the default fetch configurations](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch)

- [`logger`](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Config.html#logger-property)
  - **v2**: An object that responds to .write() (like a stream) or .log() (like the console object) in order to log information about requests.
  - **v3**: No change. More granular logs are available in v3.
- [`maxRedirects`](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Config.html#maxRedirects-property)
  - **v2**: The maximum amount of redirects to follow for a service request.
  - **v3**: **Deprecated**. SDK _does not_ follow redirects to avoid unintentional cross-region requests.
- [`maxRetries`](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Config.html#maxRetries-property)
  - **v2**: The maximum amount of retries to perform for a service request.

>   - **v3**: Changed to `maxAttempts`. See more in [v3 reference](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/interfaces/_aws_sdk_middleware_retry.retryinputconfig-2.html#maxattempts). Note that the `maxAttempt` should be `maxRetries` + 1.

Reviewer: ```suggestion
  - **v3**: Changed to `maxAttempts`. See more in [v3 reference for RetryInputConfig](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/interfaces/_aws_sdk_middleware_retry.retryinputconfig-2.html#maxattempts).
  Note that the `maxAttempt` should be `maxRetries + 1`.
```
